File: src/main/java/ru/hh/typofixer/Dictionary.java
package ru.hh.typofixer;

import com.google.common.base.Predicate;
import com.google.common.collect.ImmutableMap;
import com.google.common.collect.Maps;
import java.util.Map;

public class Dictionary {
  public static final Map<String, Integer> dict = ImmutableMap.<String, Integer>builder().
     put("hello", 1).
     put("world", 1).
     put("привет", 1).
     put("coffee", 1).
     put("java", 1).
     put("scala", 1).
     put("php", 1).
     put("программист", 1).
     put("бухгалтер", 1).
     build();

  public static Map<? extends Token, Integer> match(Iterable<? extends Token> tokens, int limit) {
    Map<Token, Integer> map = Maps.newHashMap();
    for (Token token : tokens) {
      map.put(token, token.weight(dict));
    }
    return Maps.filterValues(map, new Predicate<Integer>() {
      @Override
      public boolean apply(Integer input) {
        return input > 0;
      }
    });
  }
}

File: src/main/scala/ru/hh/typofixer/Dictionary.scala
package ru.hh.typofixer

import io.Source
import java.util

class Dictionary(val words: Array[String]) {
  def weight(word: String) = {
    val idx = util.Arrays.binarySearch(words.asInstanceOf[Array[AnyRef]], word)
    if (idx >= 0 && words(idx) == word) 10 + word.length else 0
  }
}

object Dictionary {
  def main(s:Array[String]) {
    println(prod.weight("разработчик"))
    println(prod.weight("сайтов"))
    println(prod.weight("менеджер"))
    println(prod.weight("проектов"))
    println(prod.weight("разработчик"))
  }
  def apply(words: String*) = new Dictionary(words.toArray.sorted)

  val prod = new Dictionary(Source.fromURL(getClass.getResource("/words.txt")).getLines().toArray.sorted)

  val test = Dictionary(
    "hello",
    "world",
    "привет",
    "coffee",
    "java",
    "scala",
    "php",
    "программист",
    "бухгалтер",
    "менеджер"
  )
}
